consensus/src/block_storage/block_store.rs

```diff
@@ -274,6 +283,13 @@ impl BlockStore {
             return Ok(existing_block);
         }
         let executed_block = self.execute_block(block)?;
+        // ensure local time past the block time
+        let block_time = Duration::from_micros(executed_block.timestamp_usecs());
+        let current = self.time_service.get_current_timestamp();
+        if current < block_time {
+            let wait_duration = block_time - current + Duration::from_millis(1);
+            self.time_service.sleep(wait_duration);
```

Comment:
are we going to sleep 1hour here if needs be?

Reply:
yes if 1. Quorum agrees the time is 1 hour after 2. The Round deadline is 1 hour after.
we can obviously set up counter and alert for long sleep

Reply:
when a node has time screw for 1 hour we probably don't want it to serve clients. this is no different than a down/byzantine node to me.

my principle is always serving **local_state** to clients, and replicate **remote_state** if local_state agrees with remote_state not blindly even if it's signed by quorum, this helps simplify my mental model.

the state here can be either merkle state or timestamp, think about what if a quorum signs a chain, do we execute and see if it matches the local execution or blindly take it and ignore local execution (what if it's a fork chain)? the same applies to timestamp imo.

Reply:
> if we're optimizing for this, we shouldn't wait when propose too, just propose a timestamp = parent's + 1 can minimize sleep and get proposal accepted.

Well, this would just violate the agreement that no one proposes until their timestamp follows the protocol.  My point was within the constraints of the existing timestamp properties, we should provide committed data to clients ASAP.  VASPs would want the shortest confirmation time as possible for instance.  Given the code already exists and has been running in production for a very long time without issue I'd say it's pretty well tested.

> if we're talking about seconds or more, the operator should get alert about long waiting time and fix it rather suffer it in a weird way (I mean it'd still sleep anyway)

I agree with alerting upon issues (existing metrics can be used for this alerting), but I disagree that it is necessary to unnecessarily introduce latency for clients while the issue is being debugged.  I would assume more often than not, it might be just be a situation where ntp syncing stopped or something is wrong with the hardware (e.g. something else writing to SSD or something).  Seconds matter when the expectation is to have commit latency in the seconds.

Reply:
> within the constraints of the existing timestamp properties, we should provide committed data to clients ASAP

as I said above, I disagree we should serve data that's not consistent with local state even it's signed by quorum.

**consistency weighs more than performance** imo, even full nodes should not serve the data if its local timestamp doesn't agree.

then it's a matter of how aggressively we want to fix it, if we want to minimize the latency more, we have alerts on smaller wait durations.

Reply:
Maybe it's simpler to think in terms of properties:

The existing properties we have for time are:
1) Time Monotonicity: Time is monotonically increasing in the block chain. (i.e. If H1 < H2, H1.Time < H2.Time).
2) If a block of transactions B is agreed on with timestamp T, then at least f+1 honest replicas think that T is in the past. An honest replica will only vote on a block when its own clock >= timestamp T.

There is another property that you are proposing:
3) If a block of transactions B is committed or has a QC with timestamp T, an honest validator will not serve such a block (to other validators in the case of a QC or to clients in the case of a commit) until its own clock >= timestamp T.

Reply:
that's accurate, I like this summary

Reply:
the 3 property is an extrapolation of the existing property we have:

a node only replicates (and serves) a state if it agrees with it after execution.

Reply:
^ I like this property way of describing things btw :p that'd be nice to have that in the spec

Reply:
Just glancing at the code the big delta is that instead of ending up in a stale state, which would happen if we were to wait for too long we now just sleep forever?

If my understanding is correct, I think this makes sense, we should have counters that indicate the current time before sleeping and the wake up time or the duration that way we can easily detect these kind of scenarios. It might also be useful to have an explicit state here and also have in prometheus what that state is.

In terms of implementation, with this change, I would argue we should merge secure/time and eliminate the consensus time since they effectively are the same thing now.